src/library_syscall.js

```diff
@@ -1070,11 +1070,25 @@ var SyscallsLibrary = {
       var id;
       var type;
       var name = stream.getdents[idx];
-      if (name[0] === '.') {
-        id = 1;
+      if (name === '.') {
+        id = stream.id;
         type = 4; // DT_DIR
-      } else {
-        var child = FS.lookupNode(stream.node, name);
+      }
+      else if (name === '..') {
+        id = FS.lookupPath(stream.path, { parent: true }).id;
+        type = 4; // DT_DIR
+      }
+      else {
+        try {
+          var child = FS.lookupNode(stream, name);
+        } catch (e) {
+          if (!(e instanceof FS.ErrnoError) || e.errno !== {{{ cDefine('EACCES') }}}) {
+            throw e;
+          }
+          // Access denied; skip it in the directory listing.
+          idx++;
+          continue;
+        }
```

Comment:
There seem to be several different things being fixed here. 

1. Because we only look at the first char were were treated `.foo` as `.`.
2. We were not handling `..` correctly.
3. We were listing unreadable files .. don't we want to do this?  (normally even files that are not readable do appear in directory listings).

Also the other issues in the PR seem to be:

4. Don't pass `O_DIRECTORY` through to node (so we have a test for that part).
5. `['.', '..']` were not included with noderawfs.

Is my understanding correct?

Reply:
Looking through and playing with tests, I agree with this. I don't understand why we are skipping directory entries that we can't access, and am guessing @kripken added it in d344464cc9 for some reason?

Reply:
For
> 4. Don't pass O_DIRECTORY through to node (so we have a test for that part).
are you saying we ought to add a test for that?

Reply:
If it not strictly needed as part of this change maybe lets leave it for a separate change .. and add a test for that then?